File: YotsubaEngine/Core/System/S_AGNOSTIC/SystemBuilder.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using System;
using YotsubaEngine.Core.Entity;
using YotsubaEngine.Core.System.Contract;
using YotsubaEngine.Core.YotsubaGame;
using YotsubaEngine.HighestPerformanceTypes;

namespace YotsubaEngine.Core.System.S_AGNOSTIC
{
    public class SystemBuilder : IRenderSystem
    {
        /// <summary>
        /// Almacena todas las clases definidas por el usuario
        /// que heredan de ISystem que deben ser instanciadas como sistemas 
        /// en cada escena del juego.
        /// </summary>
        public static YTB<Func<ISystem>> Systems = new YTB<Func<ISystem>>();

        /// <summary>
        /// Almacena las instancias de las clases definidas por el usuario
        /// que heredan de ISystem para la escena actual del juego.
        /// </summary>
        public YTB<ISystem> SystemsInstances = new YTB<ISystem>();

        /// <summary>
        /// Método para anadir un sistema adicional definido por el usuario.
        /// Se ejecutara en todas las escenas.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        public static void AddSystem<T>() where T : ISystem, new()
        {
            Systems.Add(() => (ISystem)new T()); 
        }

        public void InitializeSystem(EntityManager entities)
        {
            foreach (ref var system in Systems.AsSpan())
            {
                SystemsInstances.Add(system());
            }

            foreach (ref var system in SystemsInstances.AsSpan())
                system.InitializeSystem(entities);

        }

        public void SharedEntityInitialize(ref Yotsuba Entidad)
        {
            foreach (ref var system in SystemsInstances.AsSpan())
                system.SharedEntityInitialize(ref Entidad);
        }

        public void UpdateSystem(GameTime gameTime)
        {
            foreach (ref var system in SystemsInstances.AsSpan())
                system.UpdateSystem(gameTime);
        }

        public void SharedEntityForEachUpdate(ref Yotsuba Entidad, GameTime time)
        {
            foreach (ref var system in SystemsInstances.AsSpan())
                system.SharedEntityForEachUpdate(ref Entidad, time);
        }

        public void Render2D(SpriteBatch spriteBatch, GameTime gameTime)
        {
            foreach (ref var s in SystemsInstances.AsSpan())
            {
                if (s is IRenderSystem system)
                    system.Render2D(spriteBatch, gameTime);
            }
        }

        public void Render3D(GameTime gameTime)
        {
            foreach (ref var s in SystemsInstances.AsSpan())
            {
                if (s is IRenderSystem system)
                    system.Render3D(gameTime);
            }
        }
    }
}

File: Benchmarks_FSharp/Core/Systems/SystemBuilder.fs
namespace Benchmarks.Core.Systems

open System
open Benchmarks.Core.Collections
open Benchmarks.Core.Entity
open Benchmarks.Core.Scene
open Benchmarks.Core.Systems.Contract
open Benchmarks.Core.Types

type SystemBuilder() =
    inherit IRenderSystem()

    static member val Systems = YTB<Func<ISystem>>() with get, set
    member val SystemsInstances = YTB<ISystem>() with get, set

    static member AddSystem<'T when 'T :> ISystem and 'T: (new: unit -> 'T)>() =
        SystemBuilder.Systems.Add(Func<ISystem>(fun () -> new 'T() :> ISystem)) |> ignore

    override this.InitializeSystem(entities: EntityManager) =
        this.SystemsInstances.Clear()
        let systems = SystemBuilder.Systems.AsSpan()
        for i = 0 to systems.Length - 1 do
            this.SystemsInstances.Add(systems.[i].Invoke()) |> ignore

        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            instances.[i].InitializeSystem(entities)

    override this.SharedEntityInitialize(entidad: byref<Yotsuba>) =
        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            instances.[i].SharedEntityInitialize(&entidad)

    override this.UpdateSystem(gameTime: GameTime) =
        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            instances.[i].UpdateSystem(gameTime)

    override this.SharedEntityForEachUpdate(entidad: byref<Yotsuba>, time: GameTime) =
        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            instances.[i].SharedEntityForEachUpdate(&entidad, time)

    override this.Render2D(gameTime: GameTime) =
        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            match instances.[i] with
            | :? IRenderSystem as system -> system.Render2D(gameTime)
            | _ -> ()

    override this.Render3D(gameTime: GameTime) =
        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            match instances.[i] with
            | :? IRenderSystem as system -> system.Render3D(gameTime)
            | _ -> ()

    override this.Dispose() =
        let instances = this.SystemsInstances.AsSpan()
        for i = 0 to instances.Length - 1 do
            instances.[i].Dispose()
        this.SystemsInstances.Clear()
